tests/api2/test_crud_events.py

```diff
@@ -0,0 +1,135 @@
+import contextlib
+import threading
+import typing
+
+from middlewared.test.integration.assets.crypto import get_cert_params, root_certificate_authority
+from middlewared.test.integration.utils import call
+from middlewared.test.integration.utils.client import client
+
+
+def event_thread(event_endpoint: str, context: dict):
+    with client(py_exceptions=False) as c:
+        context['start_event'].set()
+        subscribe_payload = c.event_payload()
+        event = subscribe_payload['event']
+        context['event'] = event
+
+        def cb(mtype, **message):
+            if len(message) != 3 or not all(
+                k in message for k in ('id', 'msg', 'collection')
+            ) or message['collection'] != event_endpoint or message['msg'] not in (
+                'added', 'changed', 'removed'
+            ):
+                return
+
+            context['result'] = message
+            event.set()
+
+        c.subscribe(event_endpoint, cb, subscribe_payload)
+        if not event.wait(timeout=context['timeout']):
+            raise Exception('Timed out waiting for CRUD event to be generated')
+
+
+@contextlib.contextmanager
+def gather_events(event_endpoint: str, context_args: dict = None):
+    context = {
+        'result': None,
+        'event': None,
+        'start_event': threading.Event(),
+        'timeout': 60,
+        **(context_args or {})
+    }
+    thread = threading.Thread(target=event_thread, args=(event_endpoint, context))
```

Comment:
The thread is not daemon so we'll wait indefinitely for it to end on the interpreter shutdown